### PR TITLE
Add tardis

### PR DESCRIPTION
**Pull request by KevinCawley**

### CODE REVIEW

**Comment by KevinCawley** [comment]
One current issue is that tardis has several pip dependencies (sphinxcontrib-tikz, dokuwiki, dot2tex, sphinx-jsonschema, git+https://github.com/Naereen/dot2tex.git, pickle5), and I don't know how to include them into the environment setup.

**Comment by KevinCawley** [comment]
I have an open question to deal with the problem with dependencies here: https://numba.discourse.group/t/numba-integration-testing-use-conda-forge-instead-of-conda-for-texasbbq-and-pip-instillations/713

**Comment by esc** [comment]
@KevinCawley thank you for submitting this. I am very excited to see this and happy to help out where I can! 

As for `pip` dependencies. If they are specified in the `setup.py` they should be installed automagically by virtue of `python setup.py install` which you can do as an alternative to `python setup.py develop`. The other workaround would be to use adapt/override  the `install` command to inject a `pip install $PACKAGES` before running `python setup.py install` 

Using https://github.com/numba/texasbbq/blob/master/texasbbq.py#L395, for example (untested):

```
    def install(self):
        """ Custom install function for Tardis """
        if not os.path.exists(self.name):
            self.clone()
        os.chdir(self.name)
        execute("conda run --no-capture-output -n {} {}".format(self.name, "pip install ..."))
        execute("conda run --no-capture-output -n {} {}".format(self.name, "python setup.py install))
        os.chdir('../')
```

I also noticed, that you pin the  Numba version, which may not be the best idea. The integration testing setup provides a pre-built Numba from the current mainline (`master`) to test against. If Tardis is failing against this version of Numba it may indicate a bug in Numba (or Tardis, or how Tardis uses Numba). So I would recommend to drop that pinning again.

Hope that helps and please do not hesitate to get in touch with more questions!

File: .circleci/config.yml [hunk 1]
```diff
@@ -68,20 +68,18 @@ jobs:
...
     jobs:
-      - umap
-      - librosa
-      - clifford
-      - awkward
-      - sparse
-      - fastparquet
-      - datashader
-      - pandas
+      - tardis
```
Comment (esc) [inline .circleci/config.yml:90]: We will wanto re-insert these, once the tardis testing is working well.

Comment (KevinCawley) [reply .circleci/config.yml:90]: Yes, they were removed right now so we could see where tardis was failing the pipeline without waiting for ~40 minutes per run as other tests would complete.

Comment (esc) [reply .circleci/config.yml:90]: Excellent, a smart choice! 😎

**Comment by esc** [comment]
I just took another look at this and the error you are seeing is due to a Numba version mismatch. It turns out, that even though you tried to pin Numba, the TexasBBQ package took over and installed the latest nighty-build of Numba anyway. I suspect that Tardis still targets an older version (pre-refactor-2020) of Numba. You'll need to update the import statements like so:

```
In [1]: from numba import jitclass
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-507982a5b165> in <module>
----> 1 from numba import jitclass

ImportError: cannot import name 'jitclass' from 'numba' (/Users/vhaenel/git/numba/numba/__init__.py)

In [2]: from numba.experimental import jitclass
```

Give me a shout if there is anything I can do to help.

**Comment by KevinCawley** [comment]
That's the change that I currently have in a TARDIS pr as one of our files does` from numba import jitclass` instead of `from numba.experimental import jitclass`, and am waiting on a test to be updated and ran against my PR so it can be submitted and I can get a new release and run the pipeline again. 

Thank you for your input on the pip dependencies, I'll see about adding that.

**Comment by esc** [comment]
> That's the change that I currently have in a TARDIS pr as one of our files does` from numba import jitclass` instead of `from numba.experimental import jitclass`, and am waiting on a test to be updated and ran against my PR so it can be submitted and I can get a new release and run the pipeline again.

Excellent!

> Thank you for your input on the pip dependencies, I'll see about adding that.

Sure and in the long-term, maybe we can add a `pip_dependencies` abstract method to the `GitTarget` of TexasBBQ?

**Comment by KevinCawley** [comment]
That sounds like something that would be fun to do.

**Comment by esc** [comment]
> That sounds like something that would be fun to do.

Indeed and actually reasonably straight forward. 😉

**Comment by KevinCawley** [comment]
Right now for working on the texasbbq.py to add pip dependencies I have it where I am adding a pip_dependencies inside the GitTarget, and have modified the install function so that it checks to see if the pip_dependencies is not an empty string, and if not install those dependencies I will have to modify and add the function to all current targets in switchboard.py to stop the NotImplementedError. 

Is this what you had in mind, or is there a simpler approach?

**Comment by esc** [comment]
> Right now for working on the texasbbq.py to add pip dependencies I have it where I am adding a pip_dependencies inside the GitTarget, and have modified the install function so that it checks to see if the pip_dependencies is not an empty string, and if not install those dependencies I will have to modify and add the function to all current targets in switchboard.py to stop the NotImplementedError.
> 
> Is this what you had in mind, or is there a simpler approach?

This is more or less what I had in mind. I would suggest returning a list of strings however, so that "smart" users can batch `pip` installs. I.e. doing `["package1 package2", "package 3"]` will cause two `pip` invocations, namely `pip install package1 package2` and then `pip install package3`. (IIRC this is what the conda route does too, but do double check). As for the default, I would suggest returning an empty list instead of the `NotImplementedError` perhaps, so as to not break existing targets. Also `pip` dependencies are kind-of optional anyway, so that should be OK.

As a sidenote: it is conceivable that packages may want to run a complex interwoven set of `pip` and `conda` installs. I would suggest however, that this is an esoteric use-case we don't need to cater for. Anyone with this complex setup can easily just override `install` in a similar fashion as I described above to get whatever sequence is needed.

**Comment by seibert** [comment]
alternatively, `conda env create -f myenv.yaml` allows construction of an environment from a YAML file that can specify non-standard conda channels as well as pip dependencies:

https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually

**Comment by KevinCawley** [comment]
I think adding the pip_dependency option is good to have, and I will see about implementing the TARDIS env file (which it currently uses) as that will allow tardis to change dependencies in the future without a new PR having to be opened here.

**Comment by KevinCawley** [comment]
What I am doing currently with trying to install the tardis environment and all of the dependencies takes 35 minutes, where if I just install the dependencies that do math and not documentation it takes 8 minutes. I am unsure of why there is this huge time difference; is the download speed slow for the pipeline? The old Pr where I specified dependencies used 24  and the environmental file has 54.

**Comment by esc** [comment]
@KevinCawley I have no idea and it seems odd to me that this should take so long. Does it take so long locally too? Other options could be that `conda` is having issues trying to resolve the environment.

**Comment by KevinCawley** [comment]
I haven't timed it, but doing an install of tardis following the installation guide (https://tardis-sn.github.io/tardis/installation.html) does no take anywhere near 35 minutes. I would have to say that at most it takes 8 minutes for all packages. I believe conda does have problems, from what I remember it would try to do a frozen install and then go to flexible for each package, and did spend a majority on the environment.

**Comment by esc** [comment]
@KevinCawley in that case it may be better to find a minimal set of packages that can satisfy running the tardis tests related that exercise any Numba parts of the code.

**Comment by KevinCawley** [comment]
@esc  The only thing that I am afraid of is when TARDIS changes/adds some core dependencies then the tests will fail and a new PR will have to be opened, while using the environmental file allows it to pass as it will be updated automatically. For the current testing I will go back to the core dependencies until those pass and then see about using the environmental file and get it to pass as well, as there was a different error when I tried using it. The dependency error was due to a file using `jitclass`, while the new error is from a dependency usage.

**Comment by esc** [comment]
@KevinCawley yes, I was worried about this too when initially developed the package TexasBBQ and the integration testing setup for Numba. I ended up coming to the conclusion however, that there is no single way (not even a reasonably accepted subset IMHO) for Python programs to specify their dependencies. Even worse, most packages have multiple places where they define dependencies (conda packaging meta.yaml, setup.py, README.{md,rst} , sphinx documentation, pyproject.toml, requirements.txt, environment.yaml and so on....) which means it is almost never clear what one should even be looking at. So, the most practical approach is an iterative, unprincipled one: manually looking at each package, working out what the minimal set of dependencies is and then iterating on the tests until they pass. I was also worried about constant package updates and changing dependencies but since the packages covered by the integration test is fairly small and since most packages don't release daily/weekly/monthly(even) keeping the dependencies up-to-date is actually pretty trivial. I usually look at the suite once a day and broken dependencies are mostly pretty obvious. They can then be fixed with a small PR. (Note also, that I have been away last week and the week before so the current state of the integration-testing-suite is somewhat shambolic, but I intend to fix that soonish).

**Comment by esc** [comment]
Looking a bit at the error:

```
E   ImportError: libopenblas.so.0: cannot open shared object file: No such file or directory
```

And googling a bit here: https://stackoverflow.com/questions/36893382/scipy-installation-issue-getting-importerror-libopenblas-so-0-cannot-open-sha

Perhaps `sudo apt-get install libopenblas-dev` may really be a solution. However TexasBBQ is not equipped yet to deal with `apt` or other system package managers so this is something to do at the circle-ci level. We do this already for `librosa` here:

https://github.com/numba/numba-integration-testing/blob/master/.circleci/config.yml#L38

**Comment by KevinCawley** [comment]
If possible I would like to be able to also do a test against the TARDIS reference data to see if numba changes any of the scientific results, but that would involve downloading the tardis refdata in the pipeline. Am I allowed to do that in this pipeline?

**Comment by esc** [comment]
> If possible I would like to be able to also do a test against the TARDIS reference data to see if numba changes any of the scientific results, but that would involve downloading the tardis refdata in the pipeline. Am I allowed to do that in this pipeline?

Great that you managed to make the tests pass. I am not sure what `downloading the tardis refdata` means exactly. I am not aware of any download or space restrictions so my advice would be to simply attempt to download the data and see if it works.

**Comment by esc** [comment]
@KevinCawley looks really good and working -- want to make some more changes, or may I merge?

**Comment by KevinCawley** [comment]
I want to try to remove two dependencies, but other than that I am good!

**Comment by esc** [comment]
@KevinCawley sounds good, just give me the green light, when you would like it to be merged.

**Comment by KevinCawley** [comment]
@esc Ready here!

**Comment by esc** [comment]
@KevinCawley "brace for impact!" -- btw: congrats on this contribution, it is the first time someone other than myself managed to stand up a config for a project, so, well done!!! 💯 💯 💯

**Review by BigGuy573 — COMMENTED** [review]
### ```
#!/usr/bin/env python
import os
import platform
from packaging.version import parse
from texasbbq import (main,
                      execute,
                      git_ls_remote_tags,
                      git_latest_tag,
                      CondaSource,
                      GitTarget,
                      CondaTarget,
                      )
class NumbaSource(CondaSource):
    module = __name__
    @property
    def name(self):
        return "numba"
    @property
    def conda_package(self):
        return "--update-specs numba/label/dev::numba"
class UmapTests(GitTarget):
    @property
    def name(self):
        return "umap"
    @property
    def clone_url(self):
        return "https://github.com/lmcinnes/umap"
    @property
    def git_ref(self):
        return([t for t in git_ls_remote_tags(self.clone_url) if not
                t.startswith("v")][-1])
    @property
    def conda_dependencies(self):
        return ["numpy pytest nose scikit-learn pynndescent scipy pandas bokeh "
                "matplotlib datashader holoviews tensorflow scikit-image"]
    @property
    def install_command(self):
        return "pip install -e ."
    @property
    def test_command(self):
        return "pytest"
class LibrosaTests(GitTarget):
    @property
    def name(self):
        return "librosa"
    @property
    def clone_url(self):
        return "https://github.com/librosa/librosa.git"
    @property
    def git_ref(self):
        return([t for t in git_ls_remote_tags(self.clone_url) if not
                t.startswith("v")][-1])
    @property
    def conda_dependencies(self):
        return [
            "pip numpy scipy coverage scikit-learn matplotlib pytest",
        ]
    @property
    def install_command(self):
        return "conda install -c conda-forge ffmpeg pysoundfile && pip install --pre -e .[tests]"
    @property
    def test_command(self):
        return "pytest"
class CliffordTests(GitTarget):
    @property
    def name(self):
        return "clifford"
    @property
    def clone_url(self):
        return "https://github.com/pygae/clifford.git"
    @property
    def git_ref(self):
        return(git_ls_remote_tags(self.clone_url)[-1])
    @property
    def conda_dependencies(self):
        return [
            "python ipython numpy scipy pip h5py "
            "pytest pytest-benchmark",
            "-c conda-forge sparse",
        ]
    @property
    def install_command(self):
        return "python setup.py install"
    @property
    def test_command(self):
        return "pytest -v clifford/test"
class AwkwardTests(GitTarget):
    @property
    def name(self):
        return "awkward"
    @property
    def clone_url(self):
        return "https://github.com/scikit-hep/awkward-1.0.git"
    @property
    def git_ref(self):
        return(git_ls_remote_tags(self.clone_url)[-1])
    def clone(self):
        # Awkward has special needs when cloning, --recursive.
        execute(f"git clone -b {self.git_ref} {self.clone_url} "
                f"--recursive {self.name}")
    @property
    def conda_dependencies(self):
        # Awkward needs more recent compilers, support Linux and MacOSX
        operating_system = platform.uname()[0]
        if operating_system == "Linux":
            compilers = "-c conda-forge gcc_linux-64=9.3 gxx_linux-64=9.3 sysroot_linux-64=2.17"
        elif operating_system == "Darwin":
            compilers = "-c conda-forge clang_osx-64=11.0 clangxx_osx-64=11.0"
        else:
            msg = f"Operating system '{operating_system}' is not supported"
            raise NotImplementedError(msg)
        # Minimal dependencies only, 'localbuild.py'  will call on 'pip' to
        # install the rest.
        return ["python numpy make cmake", compilers]
    @property
    def install_command(self):
        # Switch -j1 is for 'make' parallellistm and --release for cmake mode.
        return "python localbuild.py -j1 --release"
    @property
    def test_command(self):
        # Switch -rfEsxX enables more verbose pytest summary.
        return "pytest -rfEsxX -v tests"
class SparseTests(GitTarget):
    @property
    def name(self):
        return "sparse"
    @property
    def clone_url(self):
        return "https://github.com/pydata/sparse.git"
    @property
    def git_ref(self):
        return git_latest_tag(self.clone_url, vprefix=False)
    @property
    def conda_dependencies(self):
        return ["pip numpy scipy"]
    @property
    def install_command(self):
        return "pip install -e .[all]"
    @property
    def test_command(self):
        return "pytest --pyargs sparse"
class FastparquetTests(GitTarget):
    @property
    def name(self):
        return "fastparquet"
    @property
    def clone_url(self):
        return "https://github.com/dask/fastparquet.git"
    @property
    def git_ref(self):
        return([t for t in git_ls_remote_tags(self.clone_url)
                if not t == "1.1"][-1])
    @property
    def conda_dependencies(self):
        return ["python numpy pandas moto cython setuptools pytest",
                # compression algos available via defaults/main
                "brotli thrift python-snappy lz4",
                # compression algos available via conda-forge
                "-c conda-forge bson python-lzo s3fs",
                # zstandard and zstd was a mess on anaconda.org at the time of
                # writing, so enable setup.py to pull in versions from PyPi
                ]
    @property
    def install_command(self):
        return "python setup.py install"
    @property
    def test_command(self):
        return "python setup.py test"
    # Tetsts need to futz about with the environment
    def test(self):
        os.environ["AWS_ACCESS_KEY_ID"] = "1111"
        os.environ["AWS_SECRET_ACCESS_KEY"] = "2222"
        super().test()
        os.environ.pop("AWS_ACCESS_KEY_ID")
        os.environ.pop("AWS_SECRET_ACCESS_KEY")
class DatashaderTests(GitTarget):
    @property
    def name(self):
        return "datashader"
    @property
    def clone_url(self):
        return "https://github.com/holoviz/datashader.git"
    @property
    def git_ref(self):
        return "v" + str(sorted([parse(t)
                   for t in git_ls_remote_tags(self.clone_url)])[-1])
    @property
    def conda_dependencies(self):
        return ["python bokeh<2.0 codecov colorcet dask[complete] "
                "datashape fastparquet flake8 nbsmoke numpy pandas pandas "
                "param pillow pyct[cmd] pytest pytest-benchmark pytest-cov "
                "scikit-image scipy toolz xarray==0.16.2 netcdf4"]
    @property
    def install_command(self):
        return "pip install -e ."
    @property
    def test_command(self):
        return "pytest datashader"
class PandasTests(CondaTarget):
    @property
    def name(self):
        return "pandas"
    @property
    def conda_package(self):
        return self.name
    @property
    def conda_dependencies(self):
        return ["python hypothesis pytest"]
    def test(self):
        # Testing pandas has special requirements.
        # See: https://github.com/pandas-dev/pandas/issues/37939
        # Obtain all the paths of the relevant test modules. This uses the
        # subprocess module to execute `conda run`, which then runs a single
        # Python command in the correct environment to import the test module
        # and print it's __file__ attribute.
        paths = []
        for module_subpath in (
            "util",
            "groupby.aggregate",
            "groupby.transform",
            "window",
            ):
            path = execute('conda run -n pandas python -c '
                           '"from pandas.tests.{} import test_numba ; '
                           'print(test_numba.__file__)"'.format(
                               module_subpath), capture=True)
            paths.append(path.decode("utf-8").strip())
        # Run pytest on all the testing modules. This again uses subprocess to
        # execute conda run to run pytest in the correct environment on the
        # correct paths of the test module.
        execute('conda run --no-capture-output -n pandas pytest '
                '{}'.format(" ".join(paths)))
class TardisTests(GitTarget):
    
    @property
    def name(self):
        return "tardis"
    
    @property
    def clone_url(self):
        return "https://github.com/tardis-sn/tardis.git"
    
    @property
    def git_ref(self):
        return (git_ls_remote_tags(self.clone_url)[-1])

    @property
    def conda_dependencies(self):
        return ["python=3", "pip", "numpy=1.19", "scipy=1.5", "pandas=1.0", "astropy=3", 
                "numba", "numexpr", "networkx", "pyyaml", "jsonschema", "pyne=0.7", 
                "pytables", "h5py", "requests", "tqdm", "beautifulsoup4", "lxml", 
                "jupyter", "notebook", "matplotlib", "graphviz", "pygraphviz", 
                "ipywidgets", "qgrid", "plotly", "pyside2", "sphinx", "nbconvert", 
                "numpydoc", "docutils>=0.16, <0.17", "nbformat", "nbsphinx", 
                "sphinx_bootstrap_theme", "sphinxcontrib-bibtex=1.0", "sphinxcontrib-apidoc", 
                "sphinx_rtd_theme", "recommonmark", "git-lfs", "pytest=5", "pytest-html", 
                "pytest-cov", "coverage", "requests", "docopt", "black", "parso=0.8"]

    @property
    def install_command(self):

    
          
            
    

          
    
    
  
        return "python setup.py develop"
    
    @property
    def test_command(self):
        return "pytest tardis --cov=tardis --cov-report=xml --cov-report=html"
if __name__ == "__main__":
    main(NumbaSource())
```

**Comment by KevinCawley** [comment]
?

**Review by BigGuy573 — COMMENTED** [review]
Enhancements-👍
I meant for this to go in my readme me file.

**Pull request merged**